### ISSUE TITLE
Enqueueing 1M elements takes too long

### ISSUE THREAD

**Issue by dmgcodevil** [typelevel/fs2#1551]
Hi,  I'm using fs2 bounded Queue and in my application and I need to enqueue ~1M elements within a single `IO`  operation.  Example:

```scala
object FS2QueuePerf extends IOApp {

  val n = 1000000

  override def run(args: List[String]): IO[ExitCode] = {

    for {
      start <- IO(System.nanoTime())
      q <- fs2.concurrent.InspectableQueue.bounded[IO, Int](n)
      _ <- (0 until n).map(i => q.enqueue1(i)).toList.sequence
      time <- IO(TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start))
      size <- q.getSize
      _ <- q.getSize.flatMap(s => IO(println(s"queue size= $size; time = $time")))
    } yield ExitCode.Success
  }
}
```

Enqueueing 100_000 elements took ~52997 ms. 1M elements ~ I wasn't  able to collect any numbers

I've tried different approaches but got the same numbers:

```scala
fs2.Stream.range(0, n).covary[IO].through(q.enqueue).compile.drain
```
~ 56966 ms

```scala
val elements: Vector[Int] = (0 until n).foldLeft(Vector.empty[Int])(_ :+ _)
val chunk = Chunk.apply(elements: _*)
fs2.Stream.chunk(chunk).covary[IO].through(q.enqueue).compile.drain
```

~ 55978 ms

I'm not very familiar with fs2 API, am I doing something wrong?

I've done some profiling and noticed that a lot of CPU time spent in two methods: `scala.collection.immutable.List.reverse()` and `scala.collection.IterableLike.sameElements()`, see  the screenshots below

![hot methods](https://i.ibb.co/FKDyv2X/hot-methods.png)
![trace](https://i.ibb.co/qxJ8nnw/trace.png)

Thanks

**Comment by ChristopherDavenport** [comment]
Remove the data constructing from the performance loop to more clearly see enqueue time.

**Comment by dmgcodevil** [comment]
@ChristopherDavenport I tried and got the same results:

```scala
object FS2QueuePerf extends IOApp {

  val n = 100000

  val elements: Vector[Int] = (0 until n).foldLeft(Vector.empty[Int])(_ :+ _)
  val chunk = Chunk.apply(elements: _*)

  override def run(args: List[String]): IO[ExitCode] = {

    for {
      start <- IO(System.nanoTime())
      q <- fs2.concurrent.InspectableQueue.bounded[IO, Int](n)
       _ <- fs2.Stream.chunk(chunk).covary[IO].through(q.enqueue).compile.drain
      time <- IO(TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start))
      size <- q.getSize
      _ <- q.getSize.flatMap(s => IO(println(s"queue size= $size; time = $time")))
    } yield ExitCode.Success
  }
}

```

**Comment by dmgcodevil** [comment]
Based  on what I'm seeing the bottleneck is here:

```scala
def publish(i: I, state: State[S]): State[S] = {
            val qs1 = strategy.publish(i, state.qs)
            if (Eq[S].eqv(state.qs, qs1)) state.copy(qs = qs1)
            else State(qs1, Set.empty)
          }
```
here: `Eq[S].eqv(state.qs, qs1)` 

Is it possible to only check the queue size?

**Comment by dmgcodevil** [comment]
I've changed the `publish` method as follows: 

```scala
          def publish(i: I, state: State[S]): State[S] = {
            val qs1 = strategy.publish(i, state.qs)
            State(qs1, Set.empty)
          }
```
Results :

```
Enqueueing 100_000 elements took ~928 ms. 
Enqueueing 1000_000 elements took ~1118 ms. 
```

**Comment by dmgcodevil** [comment]
Hm, there is one more place:

![trace_2](https://i.ibb.co/6nPrWpM/trace-2.png)

**Comment by dmgcodevil** [comment]
What's the complexity of `enqueue1` operation, `n^2` ?

**Comment by mpilquist** [comment]
cc @pchlupacek

**Comment by sergeykolbasov** [comment]
We've encountered quite the same issue, with `reverse` being too slow. Had to switch to the good ol' `ConcurrentLinkedQueue`

**Comment by SystemFw** [comment]
This is something we might look at for 3.0. The thing that I find encouraging is that the problems seem to be in the pure part of the code, rather than by hitting the limits of the `Ref + Deferred` strategy, which we need to avoid a `Sync` bound when using queues for control flow.

**Comment by entangled90** [comment]
I tried eliminating ScalaQueue in favour of Chain and the results looks better (I have a macbook air 2015):
100k elemnts -> ~ 500 millis
1M elements -> ~ 1.3/2 seconds

I just had to define this for compatibility:
 

```
implicit class ChainOps[A](private val chain : Chain[A]) extends AnyVal{
      def tail = chain.uncons match{
        case Some(_, t) => t
        case None => Chain.empty
      }
  }
```

**Comment by djspiewak** [comment]
Does 3.0 still use PubSub or is it a specialized Queue now?

**Comment by mpilquist** [comment]
Currently uses `PubSub` but that's up for debate. I created a naive queue using `Ref[F, s.c.i.Queue] + Deferred` and it didn't show any major improvements over the `PubSub` version.

**Comment by vasilmkd** [comment]
```scala
object FS2QueuePerf extends IOApp {

  val n = 1000000

  override def run(args: List[String]): IO[ExitCode] = {

    for {
      start <- IO(System.nanoTime())
      q <- fs2.concurrent.InspectableQueue.bounded[IO, Int](n)
      _ <- (0 until n).map(i => q.enqueue1(i)).toList.sequence
      time <- IO(TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start))
      size <- q.getSize
      _ <- q.getSize.flatMap(s => IO(println(s"queue size= $size; time = $time")))
    } yield ExitCode.Success
  }
}
```

This snippet, along with the `cats.effect.std.Queue` as currently implemented in https://github.com/typelevel/cats-effect/pull/1294 takes ~1500 ms to enqueue 1 000 000 elements. The queue is built on `Ref`, `Deferred` and `scala.collection.immutable.Queue`. Do note, this is Cats Effect 3.

**Comment by entangled90** [comment]
> ```scala
> ```scala
> object FS2QueuePerf extends IOApp {
> 
>   val n = 1000000
> 
>   override def run(args: List[String]): IO[ExitCode] = {
> 
>     for {
>       start <- IO(System.nanoTime())
>       q <- fs2.concurrent.InspectableQueue.bounded[IO, Int](n)
>       _ <- (0 until n).map(i => q.enqueue1(i)).toList.sequence
>       time <- IO(TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start))
>       size <- q.getSize
>       _ <- q.getSize.flatMap(s => IO(println(s"queue size= $size; time = $time")))
>     } yield ExitCode.Success
>   }
> }
> ```
> 
> 
> This snippet, along with the cats.effect.std.Queue as currently implemented in typelevel/cats-effect#1294 takes ~1500 ms to enqueue 1 000 000 elements. The queue is built on Ref, Deferred and scala.collection.immutable.Queue. Do note, this is Cats Effect 3.
> ```

From some quick test I did on latest 2.x version of cats-effect, the time it takes is similar. 
The previous comments are a bit old probably.

**Comment by diesalbla** [comment]
After reading the comments, and given that the `InspectableQueue` trait has been recently deprecated in favour of `cats.effect` queue, and the last comments by @vasilmkd and @entangled90 suggest that this problem appears solved in versions 3.0 of both cats-effect and fs2, we are not clear how to proceed.

@mpilquist Would this still be relevant for the 2.x maintenance branch, or should we close?

**Comment by diesalbla** [comment]
Based on @vasilmkd comment above, it seems this issue have been solved with recent versions of `fs2` and `cats-effect`.
